src/main/java/com/nutomic/syncthingandroid/activities/ShareActivity.java

```diff
@@ -0,0 +1,283 @@
+        String mimeType = getContentResolver().getType(uri);
+        if (mimeType != null) {
+            String displayNameColumn;
+            if (mimeType.startsWith("image/")) {
+                displayNameColumn = MediaStore.Images.ImageColumns.DISPLAY_NAME;
+            } else if (mimeType.startsWith("video/")) {
+                displayNameColumn = MediaStore.Video.VideoColumns.DISPLAY_NAME;
+
+            } else if (mimeType.startsWith("audio/")) {
+                displayNameColumn = MediaStore.Audio.AudioColumns.DISPLAY_NAME;
+
+            } else {
+                displayNameColumn = MediaStore.Files.FileColumns.DISPLAY_NAME;
+            }
+
+            Cursor cursor = null;
+            try {
+                cursor = getContentResolver().query(
+                        uri,
+                        new String[]{displayNameColumn},
+                        null,
+                        null,
+                        null
+                );
+                if (cursor != null) {
+                    cursor.moveToFirst();
+                    displayName = cursor.getString(cursor.getColumnIndex(displayNameColumn));
+                }
+            } catch (Exception e) {
+                Log.e(TAG, "Could not retrieve display name for " + uri.toString(), e);
+                // nothing else, displayName keeps null
+            } finally {
+                if (cursor != null) {
+                    cursor.close();
+                }
+            }
+        }
+        return displayName;
+    }
+
+    private class CopyFilesTask extends AsyncTask<Void, Void, Boolean> {
+        ProgressDialog progress;
+        Map<Uri, String> files;
+        Folder folder;
+        int copied = 0, ignored = 0;
+
+        CopyFilesTask(Map<Uri, String> files, Folder folder) {
+            this.files = files;
+            this.folder = folder;
+        }
+
+        protected void onPreExecute() {
+            progress = ProgressDialog.show(ShareActivity.this, null,
+                    getString(R.string.copy_progress), true);
+        }
+
+        protected Boolean doInBackground(Void... params) {
+            boolean isError = false;
+            for (Map.Entry<Uri, String> entry : files.entrySet()) {
+                InputStream inputStream = null;
+                FileOutputStream outputStream = null;
+                String outPath = folder.path + entry.getValue();
+                try {
+                    if ((new File(outPath)).isFile()) {
+                        ignored++;
+                        continue;
+                    }
+                    inputStream = getContentResolver().openInputStream(entry.getKey());
+                    outputStream = new FileOutputStream(outPath);
+                    byte[] buffer = new byte[4096];
+                    int count;
+                    while ((count = inputStream.read(buffer)) > 0) {
+                        outputStream.write(buffer, 0, count);
+                    }
```

Comment:
You should be able to replace all this with `Files.asByteSink(outFile).writeFrom(inputStream);`.